### ISSUE TITLE
False diagnostic with dyn Traits

### ISSUE THREAD

**Issue by Wardenfar** [rust-lang/rust-analyzer#13031]
This snippet of code trigger an error diagnostic from rust-analyzer.
But rustc compiles it fine.
When adding the `dyn` keyword : the diagnostic disappears.

rustc 1.63.0 (4b91a6ea7 2022-08-08)
rust-analyzer version: 0.3.1170-standalone (b6d59f2bb 2022-08-14)
edition: 2018

`[Ln 3, Col 9] expected &dyn Fn(usize), found &|{unknown}| -> ()`

```rust
fn main() {
    foo(
        &(|number| {
            // just use the parameter
            let _ = number.min(0);
        })
    );
}

fn foo(_bar: &Fn(usize)) {}
```

**Comment by lnicola** [comment]
We don't support the 2015 edition very well.

**Comment by Wardenfar** [comment]
I got this issue from the `peg` crate (proc-macro). I expanded and bisected the code (my crate is in 2021 edition).
